### PR TITLE
refactor http parser implementation in http_inspector 

### PR DESCRIPTION
**Pull request by vamshi177**
Commit Message: refactor http parser implementation in http_inspector
Additional Description: 
This update replaces the usage of the legacy http_parser in http_inspector with the `LegacyHttpParserImpl` as the first step in transitioning to the Balsa parser (part of #36433). The Parser interface is used with LegacyHttpParserImpl to preserve the current behaviour of http_parser.

Risk Level: Low (no behavioural change)
Testing: All the http_inspector tests have passed (`bazel test //test/extensions/filters/listener/http_inspector/...`)

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by alyssawilk** [comment]
/wait on CI
@wbpcode please remember to use wait tags if you're not reviewing due to CI failures, to keep this off the stale dashboard

**Comment by alyssawilk** [comment]
FWIW we had a permafail on main so this will almost certainly need a main merge now

**Comment by vamshi177** [comment]
> FWIW we had a permafail on main so this will almost certainly need a main merge now

Thanks for the info, will merge the changes from main now.

**Comment by wbpcode** [comment]
> Code coverage for source/extensions/filters/listener is lower than limit of 96.6 (96.0)
Code coverage for source/extensions/filters/listener/http_inspector is lower than limit of 96.6 (86.2)

Hi, please check the coverage issue of CI. Thanks. The code LGTM, and thanks for this contribution!

/wait

File: source/extensions/filters/listener/http_inspector/http_inspector.cc [hunk 1]
```diff
@@ -24,13 +24,12 @@ Config::Config(Stats::Scope& scope)
 const absl::string_view Filter::HTTP2_CONNECTION_PREFACE = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
 
 Filter::Filter(const ConfigSharedPtr config) : config_(config) {
-  http_parser_init(&parser_, HTTP_REQUEST);
+  // Filter for only Request Message types with NoOp callbacks
+  no_op_callbacks_ = std::make_unique<NoOpParserCallbacks>();
```
[the remark this answers is not in the document]
Comment (vamshi177) [reply source/extensions/filters/listener/http_inspector/http_inspector.cc]: yes, As this is a NoOp callback instance, there is no need for heap allocation.
updated the PR.

**Comment by vamshi177** [comment]
> > Code coverage for source/extensions/filters/listener is lower than limit of 96.6 (96.0)
> > Code coverage for source/extensions/filters/listener/http_inspector is lower than limit of 96.6 (86.2)
> 
> Hi, please check the coverage issue of CI. Thanks. The code LGTM, and thanks for this contribution!
> 
> /wait

working on adding more tests to increase coverage for http_inspector. will update the PR shortly.

**Comment by wbpcode** [comment]
@vamshi177 Thanks for your hard working. But we recently will cut a major release version, we could land the PR after that release. So, take it easy, it have enough time.

/wait

File: source/extensions/filters/listener/http_inspector/http_inspector.cc [hunk 2]
```diff
@@ -23,14 +23,12 @@ Config::Config(Stats::Scope& scope)
 
 const absl::string_view Filter::HTTP2_CONNECTION_PREFACE = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
 
-Filter::Filter(const ConfigSharedPtr config) : config_(config) {
-  http_parser_init(&parser_, HTTP_REQUEST);
+Filter::Filter(const ConfigSharedPtr config) : config_(config), no_op_callbacks_() {
+  // Filter for only Request Message types with NoOp callbacks
```
Comment (ramaraochavali) [inline source/extensions/filters/listener/http_inspector/http_inspector.cc:27]: nit: add a '.' at the end

**Review by wbpcode — APPROVED** [review]

**Pull request merged**